gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,162 @@ private String dependencyPathsOfProblematicJars(
+    }
 
-        if (stack.contains(child)) {
-          // Circular dependency check
-          if (checkedCircularDependency.add(child)) {
+    @Override
+    public int hashCode() {
+      return Objects.hash(componentResult, parent);
+    }
+  }
+
+  /**
+   * Returns mapping from Maven coordinates to their dependency paths appearing in the dependency
+   * graph
+   *
+   * @param componentResult The root project in the dependency graph
+   * @param targetCoordinatesSet The Maven coordinates to check their dependency paths
+   */
+  private ListMultimap<String, String> groupCoordinatesToDependencyPaths(
+      ResolvedComponentResult componentResult, Set<String> targetCoordinatesSet) {
+
+    ListMultimap<String, String> coordinatesToDependencyPaths =
+        MultimapBuilder.hashKeys().arrayListValues().build();
+    // No need to print the same circular dependency multiple times
+    Set<ResolvedComponentResult> checkedCircularDependency = new HashSet<>();
+
+    for (String targetCoordinates : targetCoordinatesSet) {
+      // Queue of dependnecy nodes. Each node knows its parent.
+      ArrayDeque<ResolvedComponentResultNode> queue = new ArrayDeque<>();
+      ResolvedComponentResultNode firstItem =
+          new ResolvedComponentResultNode(componentResult, null);
+      queue.add(firstItem);
+
+      // Mapping to omit duplicate dependency paths in the output. This is a mapping from Maven
+      // coordinates to dependency nodes that has the dependency of the coordinate in their direct
+      // or transitive dependencies.
+      Set<ResolvedComponentResult> nodesDependOnTarget = new HashSet<>();
+
+      while (!queue.isEmpty()) {
+        ResolvedComponentResultNode node = queue.poll();
+        ResolvedComponentResult item = node.componentResult;
+
+        ModuleVersionIdentifier identifier = item.getModuleVersion();
+        String coordinates =
+            String.format(
+                "%s:%s:%s", identifier.getGroup(), identifier.getName(), identifier.getVersion());
+        if (targetCoordinates.equals(coordinates)) {
+          String dependencyPath = node.pathFromRoot();
+          coordinatesToDependencyPaths.put(targetCoordinates, dependencyPath);
+
+          for (ResolvedComponentResultNode iter = node.parent; iter != null; iter = iter.parent) {
+            nodesDependOnTarget.add(iter.componentResult);
+          }
+        }
+
+        if (nodesDependOnTarget.contains(item)) {
+          // Do not show duplicate dependency paths. If we know that this node contains dependency
+          // having targetCoordinates in its direct/transitive dependencies, there is no need to
+          // print the dependency paths again.
+          String dependencyPath = node.pathFromRoot() + " (omitted for duplicate)";
+          coordinatesToDependencyPaths.put(targetCoordinates, dependencyPath);
+
+          for (ResolvedComponentResultNode iter = node; iter != null; iter = iter.parent) {
```

Comment:
Since these 3 lines turn up twice (for a direct and all transitive dependencies), could you extract them into a helper method? `addParents()` or `nodesDependOnTarget.addAll(getAllParents())` or some such.

Reply:
Updated to `rootToNode()` and using addAll.